### PR TITLE
feat: monitor all-sub-projecs

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team
- [ ] **Manually tested**

#### What does this PR do?

Enables --all-sub-project flag for `snyk monitor`. Data for each sub-project aka depRoot is sent to the Registry separately. 

--all-sub-project is incompatible with --project-name flag.

#### Where should the reviewer start?

monitor.ts

#### How should this be manually tested?

`snyk monitor --all-sub-project` on a gradle project with subprojs

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/BST-517
https://snyksec.atlassian.net/browse/BST-506

### CODE REVIEW

File: src/cli/commands/monitor.ts [hunk 1]
```diff
@@ -96,17 +101,20 @@ async function monitor(...args0: any[]) {
...
-      } catch (error) {
-        spinner.clear(analyzingDepsSpinnerLabel)();
-        throw error;
-      }
+      // Scan the project dependencies via a plugin
+
+      const pluginOptions = plugins.getPluginOptions(packageManager, options);
+
+      // TODO: the type should depend on multiDepRoots flag
+      const inspectResult: SingleDepRootResult|MultiDepRootsResult =
```
Comment (miiila) [inline src/cli/commands/monitor.ts]: Are we safe with removing `try/catch` here?

File: src/cli/commands/monitor.ts [hunk 2]
```diff
@@ -115,38 +123,56 @@ async function monitor(...args0: any[]) {
...
-      let leader = '';
-      if (res.org) {
-        leader = '/org/' + res.org;
+
+      // We send results from "all-sub-projects" scanning as different Monitor objects
+
+      // SingleDepRootResult is a legacy format understood by Registry, so we have to convert
+      // a MultiDepRootsResult to an array of these.
+
+      const perDepRootResults: SingleDepRootResult[] = (inspectResult as MultiDepRootsResult).depRoots
```
Comment (miiila) [inline src/cli/commands/monitor.ts]: Understand the intention, however I'd prefer less type casting here and map outside of ternary operator for readability (YMMV). I don't have a better approach, will think about it.

File: src/cli/commands/monitor.ts [hunk 3]
```diff
@@ -115,38 +123,56 @@ async function monitor(...args0: any[]) {
...
+        ? (inspectResult as MultiDepRootsResult).depRoots
+          .map((depRoot) => ({plugin: inspectResult.plugin, package: depRoot.depTree}))
+        : [inspectResult as SingleDepRootResult];
+
+      // Post the project dependencies to the Registry
+      const monOutputs: string[] = [];
+      for (const depRootDeps of perDepRootResults) {
+        // TODO(kyegupov): make snyk.monitor typed by converting src/lib/index.js to TS
+        const snykMonitor = snyk.monitor as any as (path, meta, depRootDeps) => Promise<any>;
+        const res = await snykMonitor(path, meta, depRootDeps)
```
Comment (miiila) [inline src/cli/commands/monitor.ts]: Same try/catch vs inline .catch suggestion

File: src/cli/commands/monitor.ts (hunk 1 above)
Comment (kyegupov) [reply src/cli/commands/monitor.ts]: There's a Typescript reason for that. Added a documented promiseOrCleanup function do explain the intention.

File: src/cli/commands/monitor.ts (hunk 2 above)
Comment (kyegupov) [reply src/cli/commands/monitor.ts]: Converted, made the code a bit more readable with a type guard.

**Comment by kyegupov** [comment]
(not merging because weekend :beer: )

**Comment by kyegupov** [comment]
(restarting the CI build)

**Pull request merged**